### PR TITLE
sys-devel/llvm-common: add USE=emacs for llvm modes in utils/emacs

### PR DESCRIPTION
**Pull request by alfredfo**
llvm-mir-mode only available in master and llvm-17 snapshots. https://github.com/llvm/llvm-project/commit/8cbf041ecb1c304148cd09987f4bd73a41cb51a9

@thesamesam

### CODE REVIEW

**Comment by alfredfo** [comment]
@mgorny: Using `LLVM_COMPONENTS=( llvm/utils )` instead of `LLVM_COMPONENTS=( llvm/utils/vim llvm/utils/emacs ) pulls in more junk into $S, but it makes the ebuild nicer instead of having to use "${S}../../emacs/...".

What do you prefer?

File: sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild
Comment (mgorny) [inline sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild:1]: This is a minor change, and controlled by USE flag, so let's do it in place, without losing stable keywords.

File: sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild
Comment (mgorny) [inline sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild:6]: Sort.

File: sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el
Comment (mgorny) [inline sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el:1]: Do we actually have to keep 3 files in FILESDIR? That's a horrible waste of inodes.

**Review by mgorny — CHANGES_REQUESTED** [review]

File: sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild
Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-14.0.6-r1.ebuild:1]: The keywords are kept. Followed 2. here: https://devmanual.gentoo.org/general-concepts/ebuild-revisions/index.html

But yea, will fix it in place

File: sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el
Comment (alfredfo) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el:1]: Probably not, will see what can be done instead

**Comment by alfredfo** [comment]
@mgorny: Will fix all of this tomorrow, about to sleep now. Thanks for comments

**Comment by mgorny** [comment]
No hurry, thanks!

Alternatively maybe inline these two files (but wrap them, I suppose you can do that) and use `- <<-EOF` syntax in ebuild.

File: sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el
Comment (thesamesam) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el:1]: Emacs packages tend to do this in FILESDIR but I don't know why they're not inlined as they're almost always this small.

Comment (alfredfo) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el]: done

Comment (alfredfo) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el]: @thesamesam, @mgorny: do you think something like this should be done for other Emacs packages too? The only negative I see here is if you have a bunch of ebuild versions and need to change the sitefile you can't just change the file in filesdir.

Coudldn't use elisp-site-file-install <<- EOF ... EOF here since it uses the filename when installing.

Comment (mgorny) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el]: If you don't mind me asking, are these files generally uniform? Perhaps even a better idea would be to add a simple function to the eclass that would take name, description and create the file itself.

Comment (thesamesam) [reply sys-devel/llvm-common/files/50llvm-mir-mode-gentoo.el]: They almost always look like this - two lines for search path & optionally autoload.

**Comment by mgorny** [comment]
@alfredfo, could you take a look at the function proposed in #31662 and see if it will work for you here?

**Comment by mgorny** [comment]
Oh, and let's focus on 17.0.0.9999, get that one right and then get the newest 16.x to match. There's really no point in updating < 16, since llvm-common is well, common to all slots.

**Comment by alfredfo** [comment]
@mgorny: Yes, I am waiting for it to be merged and then I'll update the PR.

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-14.0.6.ebuild:48]: The comment isn't needed, `elisp-site-file-install` will add it automatically.

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-14.0.6.ebuild:54]: Ditto.

File: sys-devel/llvm-common/llvm-common-15.0.7.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-15.0.7.ebuild:44]: Also, why do you install these files in different directories?

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-14.0.6.ebuild:48]: Will fix these things too once your elisp-make-site-file PR gets merged

Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-14.0.6.ebuild:48]: nvm, it was already merged, fixing soon

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-14.0.6.ebuild:20]:
```suggestion
	emacs? ( >=app-editors/emacs-23.1:* )
```
Please use the dependency as suggested in elisp-common documentation.

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-14.0.6.ebuild:49]: llvm-mir-mode is only available in >=llvm-17. Caused some issues due to it needing llvm-mode.el, which wasn't in loadpath due to elisp-compile only adding ./. Fixed that by setting `BYTECOMPFLAGS` so I readded it

File: sys-devel/llvm-common/llvm-common-15.0.7.ebuild
Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-15.0.7.ebuild]: yeah, fixed

File: sys-devel/llvm-common/llvm-common-15.0.7.ebuild
Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-15.0.7.ebuild]: changed that too

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-14.0.6.ebuild:34]: Sorry for noticing only now, but what is `"${S}"` doing here?

Comment (alfredfo) [reply sys-devel/llvm-common/llvm-common-14.0.6.ebuild:34]: Saw it somewhere else, just removing that works right?

**Comment by thesamesam** [comment]
@alfredfo Rebase?

File: sys-devel/llvm-common/llvm-common-14.0.6.ebuild
Comment (ulm) [inline sys-devel/llvm-common/llvm-common-14.0.6.ebuild:44]:
```suggestion
		elisp-make-site-file "${SITEFILE}" llvm
```

This needs to specify the same subdirectory as the `elisp-install` command.

**Pull request closed without merging**
